### ISSUE TITLE
Question: Is --shuffle buggy by default?

### ISSUE THREAD

**Issue by ghost** [mpv-player/mpv#4969]
```
OS: GNU / Linux Debian testing 64-bit fully updated

mpv --version
mpv 0.26.0 (C) 2000-2017 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       55.58.100
   libavcodec      57.89.100
   libavformat     57.71.100
   libswscale      4.6.100
   libavfilter     6.82.100
   libswresample   2.7.100
ffmpeg version: 3.3.4-2
```
Behavior: I play random songs from a folder that consists of 160 tracks and it played only 12 and when I tried to jump to next random track it wouldn't do so; it just stopped at 119-th track.

I have seen this behavior in the past, but not with the latest releases.

Is it a bug or a standard behavior?

**Comment by stefanos82** [comment]
After multiple times I have faced this situation, I have decided to see what is going on.

I went to `mpv/common/playlist.c` file at line `172` and copied the first `2` lines and re-created them on a demo C program to generate `160` random numbers to verify my peculiar case.

Here's the code I have used with minor tweaks to make it more obvious to me as of what numbers does it actually generate:

```c
#include <stdio.h>
#include <stdlib.h>

int main(void)
{
    int i;
    int count = 160;
    for (i = 0; i < count-1; i++) {
        int j = (int)((double)(count - i) * rand() / (RAND_MAX + 1.0));
        printf("j[%03u] -> %03d\n", i, j);
    }
    return 0;
}
```
Here's the output, pipelined with `sort` command to sort them by the third column:

```bash
bin/generic | sort -k 3,3

j[142] -> 000
j[152] -> 000
j[153] -> 000
j[155] -> 000
j[156] -> 000
j[157] -> 000
j[158] -> 000
j[140] -> 001
j[020] -> 002
j[057] -> 002
j[144] -> 002
j[071] -> 003
j[134] -> 003
j[136] -> 003
j[138] -> 003
j[145] -> 003
j[154] -> 003
j[120] -> 004
j[121] -> 004
j[056] -> 006
j[059] -> 006
j[131] -> 006
j[133] -> 006
j[150] -> 006
j[151] -> 006
j[048] -> 007
j[118] -> 007
j[148] -> 007
j[104] -> 008
j[114] -> 008
j[130] -> 008
j[143] -> 008
j[051] -> 009
j[147] -> 009
j[113] -> 010
j[149] -> 010
j[129] -> 011
j[146] -> 011
j[101] -> 012
j[115] -> 012
j[127] -> 012
j[081] -> 013
j[027] -> 014
j[109] -> 014
j[111] -> 015
j[139] -> 015
j[086] -> 016
j[132] -> 016
j[026] -> 017
j[117] -> 017
j[022] -> 018
j[122] -> 018
j[135] -> 018
j[137] -> 018
j[141] -> 018
j[018] -> 020
j[052] -> 020
j[024] -> 021
j[112] -> 021
j[107] -> 022
j[060] -> 023
j[076] -> 023
j[126] -> 023
j[128] -> 023
j[085] -> 024
j[116] -> 024
j[064] -> 025
j[088] -> 025
j[096] -> 025
j[029] -> 028
j[079] -> 028
j[100] -> 028
j[123] -> 028
j[094] -> 029
j[005] -> 030
j[044] -> 032
j[108] -> 032
j[125] -> 032
j[021] -> 033
j[082] -> 034
j[106] -> 034
j[038] -> 035
j[066] -> 035
j[124] -> 035
j[055] -> 036
j[033] -> 037
j[119] -> 037
j[072] -> 038
j[110] -> 039
j[045] -> 040
j[091] -> 040
j[008] -> 042
j[098] -> 042
j[092] -> 044
j[058] -> 046
j[042] -> 047
j[068] -> 047
j[070] -> 047
j[089] -> 048
j[105] -> 048
j[006] -> 051
j[065] -> 051
j[097] -> 051
j[078] -> 052
j[103] -> 052
j[012] -> 053
j[025] -> 054
j[080] -> 055
j[099] -> 055
j[102] -> 055
j[050] -> 057
j[093] -> 057
j[069] -> 060
j[095] -> 060
j[036] -> 061
j[075] -> 061
j[077] -> 061
j[001] -> 062
j[040] -> 063
j[084] -> 063
j[035] -> 065
j[087] -> 065
j[030] -> 066
j[090] -> 066
j[083] -> 067
j[053] -> 070
j[067] -> 070
j[010] -> 071
j[013] -> 075
j[032] -> 078
j[034] -> 080
j[074] -> 080
j[073] -> 081
j[063] -> 082
j[009] -> 083
j[019] -> 085
j[062] -> 088
j[016] -> 091
j[041] -> 091
j[046] -> 092
j[011] -> 093
j[039] -> 093
j[054] -> 094
j[061] -> 096
j[017] -> 102
j[047] -> 103
j[043] -> 104
j[049] -> 105
j[031] -> 108
j[023] -> 110
j[007] -> 117
j[037] -> 119
j[002] -> 123
j[003] -> 125
j[028] -> 131
j[015] -> 132
j[000] -> 134
j[014] -> 139
j[004] -> 142
```
As you can see, we don't get the best results out of random mechanism; lots of values are generated and if we isolate those unique values, we can see that out of `160` (which in reality it's `159` based on `for()` loop that takes `count - 1`), only `84` (in my case) are the actual generated track numbers:

Before uniq use:
```bash
bin/generic | sort -k 3,3 | wc -l
159
```

After uniq use:
```bash
bin/generic | sort -uk 3,3 | wc -l
84
```
In other words, command `mpv --shuffle '/my/directory/with/songs/` will play only `84` songs out of `160` which is of course both unfair and annoying.

Is this the expected behavior the dev team had originally in mind?

Please advice, thank you.

**Comment by SteveJobzniak** [comment]
Yikes! That sounds very broken, doesn't it? @wm4 @avih @jeeb

**Comment by Argon-** [comment]
See this commit message: https://github.com/mpv-player/mpv/commit/b9c48ca8f33e63549f51edb08bd50cc6cc8badbb
So I guess either the implementation is incorrect or the stated conditions are not satisfied?

**Comment by stefanos82** [comment]
@Argon- as you can see

```c
for (int n = 0; n < count - 1; n++) {
        int j = (int)((double)(count - n) * rand() / (RAND_MAX + 1.0));
        MPSWAP(struct playlist_entry *, arr[n], arr[n + j]);
}
```
is indeed very inaccurate for the simplest reason: it tries to swap a value that does not exist.

What do I mean by that? 

Well, `count` in my case it's `160`, minus `1` it's of course `159`. When I will reach, for example element number `150` and the randomly generated value that got assigned in `j` is `110`, then `MPSWAP` tries to swap two positions that one simply does not exist, that is `arr[150]` with `arr[150+110]` (remember, j is `110`) which is of course `260` which again leads to buffer overflow!

...unless I did not understand how the code actually works.

@SteveJobzniak your iteration suggestion does not simply apply as the loop executes based on the the number of tracks `playlist_count` assigned to `count`.

**Comment by SteveJobzniak** [comment]
```
// imagine count is 150
for (int n = 0; n < count - 1; n++) {
        // n = 103

        // j = count (150) - n (103) = 47
        // j = 47 * rand() (so a value from 0 to 47 is generated, I think... let's say 20
        int j = (int)((double)(count - n) * rand() / (RAND_MAX + 1.0));

        // swap: n (103) and n+j (103+20) = 123
        MPSWAP(struct playlist_entry *, arr[n], arr[n + j]);
}
```

The algorithm looks fine to me.

What do you think @Argon- ?

**Comment by SteveJobzniak** [comment]
Btw a more natural Fisher Yates shuffle is structured differently. Start the loop at the max element and shuffle remaining elements. Less bug-prone that way. Like this:

https://github.com/SteveJobzniak/mpv-tools/blob/d8eeea4bcdf9e3538309bb7f33f37cde82980c77/scripts/modules.js/MicroUtils.js#L113-L128

**Comment by stefanos82** [comment]
@SteveJobzniak thank you very much for the thorough explanation. 

I have learned something new today that I honestly didn't know until now.

For years I would use `rand() % <n>` for ranged values in place of `<n> * rand()` and it would work as I expected it to.

It seems I have completely misunderstood how the aforementioned code actually works.

My apologies if I have offended anyone.

Last night I tried a simple approach and produced the desired results:

```c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define COUNT 160

#define RANDOM(count,index) \
    ((int)((double)(count-index) * rand() / (RAND_MAX + 1.0)))

#define MYSWAP(x,y)                 \
    do {                            \
        __typeof__ (x) _tmp = (x);  \
        x = (y);                    \
        y = _tmp;                   \
    } while(0)

int main(void)
{
    int i, j;
    int arr[COUNT] = {0};

    /* I use this function so I can generate different results 
        every time this program gets executed.
    */
    srand(time(NULL));

    for (i = 0; i < COUNT; i++) {
        arr[i] = i;
    }

    for (i = 0; i < COUNT - 1; i++) {
        j = RANDOM(COUNT,i);
        MYSWAP(arr[i], arr[i+j]);
    }

    for (i = 0; i < COUNT; i++) {
        printf("[%03d]: %03d\n", i, arr[i]);
    }

    return 0;
}
```

Indeed the code works as expected.

Now the big question remains: what causes `mpv` to randomly stop when flag `--shuffle` is used?

**Comment by SteveJobzniak** [comment]
I agree that this line could be clearer:

`int j = (int)((double)(count - n) * rand() / (RAND_MAX + 1.0));`

Rewritten something like this would make it clear that it's generating a random number, dividing it by the maximum possible random number +1 (to get a float from 0.0 to 0.999999999<infinity> but never 1):

`int j = (int)((double)(count - n) * (rand() / (RAND_MAX + 1.0)));`

That in turn means that whatever is on the left side is being multiplied by anywhere from `*0` to `*0.9999999`, (such as "47.8333") and then the result is forcibly cast to an integer which throws away the decimals (such as "47"). And voila, you get a valid integer offset.

So yeah it seems the algorithm is fine and the shuffle problem is elsewhere. I am not involved in this project so I can't help you with that part. I was just trying to help regarding the shuffle code. Good luck :) And try `#mpv` on freenode IRC chat to see if anyone else there has encountered this problem.

**Comment by aleksejrs** [comment]
For a few days now (probably not before July 28th, when I updated VLC to a buggier one, and not every day and not all the time), I've been using mpv like this:
`mpv %c --load-scripts=no --shuffle`, where %c is a directory tree with thousands of files, mostly video files (76-79 hours) accompanied with JPEG images and non-playable files.

Three days in a row, I met the same song (or a couple of songs, or maybe similar songs or different versions of the same song) at least once. A bit strange, but not a big problem.

Today, when I started it — here I may have skipped a couple of times more than I mention, without noticing; also there are usually 2 or 3 still images between each pair of songs, which take a few moments for MPV to show:
* Song A I decided not to listen to. I pressed ">" in a few seconds.
* (pictures?)
* Song B I didn't want to listen to. I pressed ">" almost immediately.
* (pictures)
* Song B again. I pressed ">" almost immediately.
* (pictures?)
* Song B again. I pressed ">" almost immediately.
Various songs followed.

Debian GNU/Linux testing, not fully updated (right now at least 91 packages are not, all but UUID and messaging services seem completely irrelevant).
```
mpv 0.29.0 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.14.100
   libavcodec      58.18.100
   libavformat     58.12.100
   libswscale      5.1.100
   libavfilter     7.16.100
   libswresample   3.1.100
ffmpeg version: 4.0.2-1
```

**Comment by stefanos82** [comment]
As of today, 21-06-2019, I have noticed a weird behavior that relates to shuffling.

I use 
```
mpv 0.29.1 Copyright © 2000-2018 mpv/MPlayer/mplayer2 projects
 built on UNKNOWN
ffmpeg library versions:
   libavutil       56.14.100 (runtime 56.22.100)
   libavcodec      58.18.100 (runtime 58.35.100)
   libavformat     58.12.100 (runtime 58.20.100)
   libswscale      5.1.100 (runtime 5.3.100)
   libavfilter     7.16.100 (runtime 7.40.101)
   libswresample   3.1.100 (runtime 3.3.100)
ffmpeg version: 4.1.3-1
```
and the issue goes like this: I use `mpv --shuffle /my/directory/of/songs/` and since I have `save-position-on-quit` inside `$HOME/.config/mpv/mpv.conf`, when I press `q`, close my terminal, re-open it, and re-play aforementioned command to resume my playlist, from time to time it re-shuffles to a newly played song and does not resume from the last played song!

It does not always happen, but it *does* happen and leaves me speechless.

**Comment by qmega** [comment]
That seems like a separate issue; consider opening a new one.

My first guess would be that mpv is not exiting successfully on the run before it fails to resume (crashing, being killed, maybe just exiting with an error). Resume information is deleted when it's used and isn't written again until exit, so if mpv dies unexpectedly you'll end up with no resume info saved, and the next run will start fresh.

If that doesn't help, I think we'd need logs from a run that didn't resume and the last run before that to debug.

You could use something like this to play so that you'd always have the last two logs:
```
mv mpv.log mpv-last.log; mpv --log-file=mpv.log ...
```

**Comment by stefanos82** [comment]
OK, I will keep an eye on it and log its behavior when it repeats it again, because as I have said above it does not always happen, but it does happen at random times completely unexpectedly.

**Comment by stefanos82** [comment]
After all this time, the issue remains the same.

I was wondering; is the value zero considered a valid input for shuffling or not? Because if that is the case, then I would assume is the root of the problem.

**Comment by stefanos82** [comment]
@qmega it seems there's somewhere a bug in core, because I run `mpv /home/stefanos/Music/Various/` and after a week, it decided today to unexpectedly stop playback without any warning.

As you can see, I'm not using `--shuffle` and behaved the same as my original ticket.

I'm currently monitoring its behavior with `--log-file=mpv.log` as you have suggested to see what causes this rather strange behavior.

**Comment by Traneptora** [comment]
Fixed by 582f3f7cc01f81345df5c20a012b1f47587e6a97.

**Comment by aleksejrs** [comment]
>  For a few days now (probably not before July 28th, when I updated VLC to a buggier one, and not every day and not all the time), I've been using mpv like this:
mpv %c --load-scripts=no --shuffle,

I may have been calling it with %f and multiple directories, and hitting #7922.
